### PR TITLE
Fix typescript errors in Vue files, fix regression in "Recent Commits" chart

### PR DESCRIPTION
**Pull request by silverwind**
- Fix all typescript errors in `.vue` files
- Fix regression from https://github.com/go-gitea/gitea/pull/32329 where "Recent Commits" chart would not render.

### CODE REVIEW

File: web_src/js/components/RepoCodeFrequency.vue [hunk 1]
```diff
@@ -108,10 +111,9 @@ function toGraphData(data) {
   };
 }
 
-const options = {
+const options: ChartOptions<'line'> = {
   responsive: true,
   maintainAspectRatio: false,
-  animation: true,
```
Comment (silverwind) [inline web_src/js/components/RepoCodeFrequency.vue:114]: `animation: true` is not a valid option, only `false` or `object` are.

File: web_src/js/components/RepoRecentCommits.vue [hunk 2]
```diff
@@ -61,11 +63,11 @@ async function fetchGraphData() {
...
     } while (response.status === 202);
     if (response.ok) {
-      const data = await response.json();
-      const start = Object.values(data)[0].week;
+      const dayDataObj: DayDataObject = await response.json();
+      const start = Object.values(dayDataObj)[0].week;
       const end = firstStartDateAfterDate(new Date());
       const startDays = startDaysBetween(start, end);
-      data.value = fillEmptyStartDaysWithZeroes(startDays, data).slice(-52);
+      data.value = fillEmptyStartDaysWithZeroes(startDays, dayDataObj);
```
Comment (silverwind) [inline web_src/js/components/RepoRecentCommits.vue:70]: Bug was here, local variable `data` was shadowing top-level `data`.

File: web_src/js/components/RepoRecentCommits.vue [hunk 3]
```diff
@@ -61,11 +63,11 @@ async function fetchGraphData() {
...
     } while (response.status === 202);
     if (response.ok) {
-      const data = await response.json();
-      const start = Object.values(data)[0].week;
+      const dayDataObj: DayDataObject = await response.json();
+      const start = Object.values(dayDataObj)[0].week;
       const end = firstStartDateAfterDate(new Date());
       const startDays = startDaysBetween(start, end);
-      data.value = fillEmptyStartDaysWithZeroes(startDays, data).slice(-52);
+      data.value = fillEmptyStartDaysWithZeroes(startDays, dayDataObj).slice(-52);
```
Comment (silverwind) [inline web_src/js/components/RepoRecentCommits.vue:70]: Bug was here, local variable `data` was shadowing top-level `data`.

File: web_src/js/components/RepoActionView.vue [hunk 4]
```diff
@@ -269,19 +270,19 @@ const sfc = {
...
+    isExpandable(status: ActionsStatus) {
       return ['success', 'running', 'failure', 'cancelled'].includes(status);
     },
 
     closeDropdown() {
       if (this.menuVisible) this.menuVisible = false;
     },
 
-    toggleTimeDisplay(type) {
+    toggleTimeDisplay(type: 'seconds' | 'stamp') {
```
Comment (silverwind) [inline web_src/js/components/RepoActionView.vue]:
```suggestion
    toggleTimeDisplay(type: string) {
```

File: web_src/js/components/RepoContributors.vue [hunk 5]
```diff
@@ -225,7 +229,7 @@ export default {
       }
     },
 
-    getOptions(type) {
+    getOptions(type: 'main' | 'contributor'): ChartOptions<'line'> {
```
Comment (silverwind) [inline web_src/js/components/RepoContributors.vue]:
```suggestion
    getOptions(type: string): ChartOptions<'line'> {
```

File: templates/repo/commit_status.tmpl [hunk 6]
```diff
@@ -1,4 +1,4 @@
-<!-- make sure this matches the color logic in web_src/js/components/DashboardRepoList.vue -->
+<!-- make sure this matches the color logic in web_src/js/data.ts -->
```
Comment (wxiaoguang) [inline templates/repo/commit_status.tmpl]:
```suggestion
<!-- make sure this matches the color logic of "commitStatus" in web_src/js/data.ts -->
```

File: web_src/js/data.ts [hunk 7]
```diff
@@ -0,0 +1,10 @@
+import type {CommitStatusMap} from './types.ts';
+
+// make sure this matches templates/repo/commit_status.tmpl
+export const commitStatus: CommitStatusMap = {
```
Comment (wxiaoguang) [inline web_src/js/data.ts]: Actually, if it is only used once, I do not see why it should be put in a common package.

Another problem is, if you'd like to put it in a common package, then it should have a clearer name, but not `commitStatus`

File: web_src/js/globals.d.ts [hunk 8]
```diff
@@ -63,3 +63,7 @@ interface Window {
   hcaptcha: any,
   codeEditors: any[],
 }
+
+interface HTMLElement {
+  _jobLogTime: number,
```
Comment (wxiaoguang) [inline web_src/js/globals.d.ts]: It doesn't seem right to be in "global"

It is only privately used

File: web_src/js/types.ts [hunk 9]
```diff
@@ -59,3 +59,20 @@ export type FomanticInitFunction = {
...
+export type CommitStatusMap = {
+  [status in CommitStatus]: {
+    name: string,
+    color: string,
+  };
+};
+
+export type ActionsStatus = 'unknown' | 'waiting' | 'running' | 'success' | 'failure' | 'cancelled' | 'skipped' | 'blocked';
+
+export type ActionsStepLogLine = {
```
Comment (wxiaoguang) [inline web_src/js/types.ts]: I am sure it is only privately used, is it really worth to move everything into a global "types" package?

File: web_src/js/globals.d.ts (hunk 8 above)
Comment (delvh) [inline web_src/js/globals.d.ts]: Erm… this looks like quite a pollution of the global type namespace?
After all, only some elements within `Actions` will actually have this attribute, right?

File: web_src/js/components/RepoContributors.vue [hunk 10]
```diff
@@ -65,10 +69,10 @@ export default {
   data: () => ({
     isLoading: false,
     errorText: '',
-    totalStats: {},
-    sortedContributors: {},
+    totalStats: {} as Record<string, any>,
+    sortedContributors: {} as Record<string, any>,
     type: 'commits',
-    contributorsStats: [],
```
Comment (delvh) [inline web_src/js/components/RepoContributors.vue:71]: How were you able to change the type from array to object without changing anything related to this object?
Was this documented incorrectly before or what happened?

File: web_src/js/globals.d.ts (hunk 8 above)
Comment (wxiaoguang) [reply web_src/js/globals.d.ts]: I agree with you 😄 

https://github.com/go-gitea/gitea/pull/32649#discussion_r1859404265

Comment (silverwind) [reply web_src/js/globals.d.ts]: Hmm, the only other option would be `as` I guess, but I find this cleaner.

File: web_src/js/types.ts (hunk 9 above)
Comment (silverwind) [reply web_src/js/types.ts]: This one likely is only for the actions log yes. The `ActionsStatus`, I could see being used elsewhere too.

File: web_src/js/components/RepoContributors.vue (hunk 10 above)
Comment (silverwind) [reply web_src/js/components/RepoContributors.vue:71]: This was an incorrect default value. It's being assigned object after the fetch. The code itself runs `Object.keys` on the value before, but the output of those is `[]` in both cases, so it's compatible.

File: web_src/js/components/RepoContributors.vue [hunk 11]
```diff
@@ -211,9 +215,9 @@ export default {
       };
     },
 
-    updateOtherCharts(event, reset) {
-      const minVal = event.chart.options.scales.x.min;
-      const maxVal = event.chart.options.scales.x.max;
+    updateOtherCharts({chart}: {chart: Chart}, reset?: boolean) {
```
Comment (silverwind) [inline web_src/js/components/RepoContributors.vue]:
```suggestion
    updateOtherCharts({chart}: {chart: Chart}, reset?: boolean = false) {
```

File: web_src/js/globals.d.ts (hunk 8 above)
Comment (wxiaoguang) [reply web_src/js/globals.d.ts]: Maybe the line `      div._jobLogTime = line.timestamp;` could removed (the only usage)

No read to the value.

Comment (silverwind) [reply web_src/js/globals.d.ts]: Seems you are right, it's a pointless assignment.

File: web_src/js/components/RepoContributors.vue [hunk 12]
```diff
@@ -238,6 +242,7 @@ export default {
             position: 'top',
             align: 'center',
           },
+          // @ts-expect-error: bug in chartjs types
```
Comment (silverwind) [inline web_src/js/components/RepoContributors.vue]:
```suggestion
          // @ts-expect-error: bug in chart.js types
```

File: web_src/js/components/RepoActionView.vue
Comment (wxiaoguang) [inline web_src/js/components/RepoActionView.vue:1]: All the changes in this file have been taken into  Refactor RepoActionView.vue, add group support #32713

File: web_src/js/globals.d.ts (hunk 8 above)
Comment (wxiaoguang) [reply web_src/js/globals.d.ts]: This line has been removed in Refactor RepoActionView.vue, add group support #32713

**Comment by silverwind** [comment]
Will pick this up again in a few days.

**Comment by wxiaoguang** [comment]
I managed to merge from main branch (189722b5e96d2e21d9102302a2d7fb6a61fae300) to resolved the conflicts caused by my PRs.

**Comment by wxiaoguang** [comment]
And since it contains a bug fix, add the 1.23 milestone

File: tsconfig.json [hunk 13]
```diff
@@ -27,6 +27,7 @@
     "noPropertyAccessFromIndexSignature": false,
     "exactOptionalPropertyTypes": false,
     "sourceMap": true,
+    "tsBuildInfoFile": "./node_modules/.cache/tsbuild",
```
Comment (silverwind) [inline tsconfig.json]:
```suggestion
```

**Comment by silverwind** [comment]
Thanks, looking good. This is ready.

**Review by wxiaoguang — APPROVED** [review]
Thank you for confirming, then LGTM

**Review by lunny — APPROVED** [review]

**Pull request merged**